pkg/sac/allow_fixed_scopes_checker_core.go

```diff
@@ -52,6 +53,81 @@ func (c allowFixedScopesCheckerCore) SubScopeChecker(key ScopeKey) ScopeCheckerC
 }
 
 func (c allowFixedScopesCheckerCore) EffectiveAccessScope(resource permissions.ResourceWithAccess) (*effectiveaccessscope.ScopeTree, error) {
-	// TODO(ROX-9537): Implement it
-	panic("Implement me: ROX-9537")
+	if len(c) == 0 {
+		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
+	}
+	for key := range c[0] {
```

Comment:
Using an array with magic indexes might be confusing: why c[0] and not c[5]? Could there be a structure with properly named fields instead, or constants for those indexes?

Reply:
I tend to agree with that approach.
Ideally, the indexes in that list should map as follows:
- 0 - access mode (None, Read, Write)
- 1 - target resource
- 2 - cluster IDs in scope
- 3 - namespace names in scope
This is the case in all production code calls, but until PR 1106 (https://github.com/stackrox/stackrox/pull/1106/) is merged, some test uses start at different levels.
In the pull request at hand, the point is to extract an effective access scope from the list of keysets. Redesigning the object would involve new constructors and a significantly larger pull request (~100 files impacted). Follow-up with JIRA ticket https://issues.redhat.com/browse/ROX-9981

Reply:
Please add a todo here

Reply:
Maybe we can have a local const that map 0 and 1 as they are only used here?

Reply:
`next()` function is a good idea, it removes some of the magic index logic and makes the code more readable.

I have multiple reasons not to integrate the other two functions.
- One should be solved by PR 1106, where the occurrences of the following pattern are replaced by a different object. The pattern (`ScopeKey` sets not starting at AccessMode level) is still possible in the code until this scope injection utility is reworked to impose more constraints on the way it can be used (one option would be to have a constructor for each valid level of scoping, ie `(AccessMode,Resource)`, `(AccessMode,Resource,Cluster)` or `(AccessMode,Resource,Cluster,Namespace)`):
```
sac.OneStepSCC{
	sac.AccessModeScopeKey(storage.Access_READ_ACCESS): sac.OneStepSCC{
		sac.ResourceScopeKey(resources.Cluster.GetResource()): sac.AllowAllAccessScopeChecker(),
		sac.ResourceScopeKey(resources.Node.GetResource()):    sac.AllowAllAccessScopeChecker(),
		sac.ResourceScopeKey(resources.Deployment.GetResource()): sac.AllowFixedScopes(
			sac.ClusterScopeKeys(clusterID),
			sac.NamespaceScopeKeys(namespace2),
		),
	},
}
```
- Another one is that `c[0]` on the result of `c.next()` is not a set of `AccessModeScopeKey` objects anymore.

Removal of the magic 0 index can be achieved through the following function:
```
func (c allowFixedScopesCheckerCore) topLevelKeys() scopeKeySet {
	return c[0]
}
```